### PR TITLE
[POC] How to handle CUD actions of Manufacturer BO page?

### PR DESCRIPTION
**Pull request by mickaelandrieu**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | -
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| How to test?  | Don't review it please

## What the hell are you doing?

We have to deal with persistence of legacy object models in the modern controllers.
Obviously, we don't want to be dependent of ObjectModel instance in controllers or even in services injected in it. So this is my C(Q)RS proposal.

> Note 1: we don't use ES for now: as the time is limited I don't think we will have time to do it in 1.7.x if we want to stick with deadlines.

> Note 2: I'm not sure we will need to implement a lot of Queries as the Grid component have his own system to retrieve data: for most of the pages, only implementing Command for C~~R~~UD actions will be enough.

### I don't understand what you're doing here!

Chill my friend :)

The idea here is to rely on Commands (that could be called using events later ;) ) in Controllers.

For instance, let's imagine a `ManufacturerController`:

```php

use PrestaShop\PrestaShop\Adapter\Manufacturer\Model\Manufacturer;
use PrestaShop\PrestaShop\Adapter\Manufacturer\Command\CreateManufacturerCommand;

/**
 * ...
 */
class ManufacturerController extends FrameworkAdminController
{
     /**
      * Most of Queries are already managed by the Grid Component.
      */
     public function indexAction(){ ... };
     public function searchAction(){ ... };

     /**
      * Create a new Manufacturer action.
      */
     public function createAction(Request $request)
     {
         $createManufacturerForm = ... ; // get the form
         $createManufacturerForm->handleRequest($request);
         
         if ($createManufacturerForm->isSubmitted()) {
             $data = $createManufacturerForm->getData();
            
             return new CreateManufacturerCommand(Manufacturer::createFromData($data));
         }
     }
}
```

We probably need a CommandBus here, and a way to retrieve errors the same way we do "Configuration" pages.

Also - contrary to all docs about CQRS I've found - our main issue here is that we have no Domain and dynamic object models (any module can add/remove any attribute of the existing tables and add new ones!). This make the validation of objects really hard to do, and type hinting and implementation really tricky to validate a "model".

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9314)

### CODE REVIEW

File: src/Core/Manufacturer/ManufacturerInterface.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
+<?php
```
Comment (kpodemski) [inline src/Core/Manufacturer/ManufacturerInterface.php:1]: @mickaelandrieu do we really need an interface here if that's a case only for one object? isn't it overkill?

File: src/Adapter/Manufacturer/Repository/ManufacturerRepository.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    public function __construct(Connection $connection, $tablePrefix)
+    {
+        $this->connection = $connection;
+        $this->tablePrefix = $tablePrefix;
+    }
+
+    public function create(ManufacturerInterface $manufacturer)
+    {
+        $legacyManufacturer = new LegacyManufacturer($manufacturer->getId(), $manufacturer->getIdLang());
+        $legacyManufacturer->save();
```
Comment (mickaelandrieu) [inline src/Adapter/Manufacturer/Repository/ManufacturerRepository.php:63]: not that if we haven't Active Record already, we'd probably have to move this responsibility to another class: this is not "job" of a repository to persist data in the database.

File: src/Core/Manufacturer/ManufacturerInterface.php (hunk 1 above)
Comment (mickaelandrieu) [reply src/Core/Manufacturer/ManufacturerInterface.php:1]: it's never overkilled when you need to uncouple classes: it's not about the quantity of code, it's about describing the direction of dependencies.

Implementation = hard bounding
Contract = Dependency Inversion

**Comment by mickaelandrieu** [comment]
Closed in favor of https://github.com/PrestaShop/PrestaShop/pull/9370

**Pull request closed without merging**